File: server/internal/api/history.go
package api

import (
	"context"
	"encoding/json"
	"net/http"
	"time"
)

const (
	// historyWindow is how far back the summary looks.
	//
	// Long enough that a route's quiet day does not move its median, short
	// enough that a timetable change from a month ago is not still being
	// averaged in with how the route runs today. Aggregates are kept for 90
	// days; this is the slice of them that describes the present.
	historyWindow = 28 * 24 * time.Hour

	// historyTTL bounds how often the medians are recomputed. The rollup only
	// produces new observations every rollup.Interval, so recomputing faster
	// than that reads the same rows to reach the same answer.
	historyTTL = 30 * time.Minute
	// A summary of four weeks does not go stale in an hour. Serving a slightly
	// old one beats failing, because the client's fallback is a hardcoded guess.
	historyMaxStale = 24 * time.Hour

	historyKey = "history"
)

// handleHistory serves what the recorded history says about how the network
// actually runs: measured headways, measured ride times between stops, and
// measured lateness against the published timetable.
//
// This is the one endpoint that serves something RTL does not publish. Every
// other route here is a cache or a reshaping of upstream data; this is the
// product of the recorder, and it exists because four of these routes publish
// no timetable at all and the rest publish one the buses do not keep to.
//
// It answers 503 rather than an empty summary when there is no store, so a
// client can tell "this server cannot measure" from "this route has not been
// measured yet" — the first is a reason to keep its own assumptions, the second
// is a reason to keep asking.
func (s *Server) handleHistory(w http.ResponseWriter, r *http.Request) {
	if s.readDB == nil {
		writeJSON(w, r, http.StatusServiceUnavailable, "no-store",
			map[string]string{"error": "history is not available on this server"})
		return
	}
	payload, _, err := s.history.Get(r.Context(), historyKey, s.loadHistory)
	if err != nil {
		s.fail(w, r, "history", err)
		return
	}
	// Cacheable for a while: it changes on the rollup's half-hourly cadence, and
	// a client that holds it for the length of a session is holding something
	// still true.
	writeRaw(w, r, "public, max-age=600, stale-while-revalidate=3600", payload)
}

func (s *Server) loadHistory(ctx context.Context) (json.RawMessage, error) {
	now := time.Now()
	summary, err := s.readDB.History(ctx, now.Add(-historyWindow).UnixMilli(), now.UnixMilli())
	if err != nil {
		return nil, err
	}
	return json.Marshal(summary)
}

File: src/api/history.ts
/**
 * What the backend's recorder has measured about how the network actually runs.
 *
 * This is the one thing the server knows that RTL does not publish. Everything
 * else the backend serves is a cache or a reshaping of upstream data; this is
 * the product of watching buses for weeks, and it exists because four of these
 * routes publish no timetable at all and the rest publish one the buses keep to
 * only approximately.
 *
 * Entirely optional. There is no fallback path to RTL because RTL has nothing to
 * fall back to — when this is unavailable the planner keeps the assumptions it
 * has always used, which is exactly what it did before the recorder existed.
 */
import { fetchFromBackend } from './backend';

/** One route's measured behaviour. Every field may be absent. */
export interface RouteHistory {
  /**
   * Median wait in minutes, absent when too little was observed to say. Missing
   * is not zero: it means "no measurement", and the caller keeps its assumption.
   */
  headwayMin?: number;
  headwaySamples: number;
  /**
   * True when the wait had to be measured as one bus coming round again,
   * because the route is worked by a single vehicle and no two consecutive
   * arrivals are different buses. Still the wait a rider has — R12 is the case
   * this exists for — but worth being able to tell apart.
   */
  headwayIsLap?: boolean;
  /** Median wait per hour of the Malé day, keyed "0".."23". */
  headwayByHour?: Record<string, number>;
  /**
   * Median signed minutes late against the route's own published timetable, per
   * hour. Positive is late. Absent for the routes that publish no timetable.
   */
  latenessByHour?: Record<string, number>;
  latenessSamples: number;
  /** Median ride in seconds between adjacent stops, keyed "fromStop>toStop". */
  segmentSecs?: Record<string, number>;
  /**
   * Median seconds to ride from one stop to another, keyed "fromStop>toStop",
   * for every ordered pair a bus was observed covering in one pass.
   *
   * Measured whole rather than summed from the legs between, so it carries what
   * a rider actually sits through. This is what a journey's ride time is taken
   * from where it exists; `segmentSecs` remains the per-leg figure.
   */
  rideSecs?: Record<string, number>;
  /**
   * The same medians resolved by hour of the Malé day, keyed "fromStop>toStop"
   * then "0".."23". Sparse: only the buckets with enough observations appear,
   * and `segmentSecs` above is what a caller falls back to for the rest.
   */
  segmentSecsByHour?: Record<string, Record<string, number>>;
}

export interface HistorySummary {
  generatedAtMs: number;
  fromMs: number;
  /** Keyed by route code ("133"), not route number ("R1"). */
  routes: Record<string, RouteHistory>;
}

/**
 * Fetches the measured summary, or null when there is none to be had.
 *
 * Null covers every failure mode there is — no backend configured, a server
 * without a store, a breaker holding it aside, a request that timed out — and
 * they all mean the same thing to the caller: plan with the assumptions.
 */
export async function fetchHistory(signal?: AbortSignal): Promise<HistorySummary | null> {
  const summary = await fetchFromBackend<HistorySummary>('/v1/history', signal);
  // A server that answered but has measured nothing is not an error, and is
  // still worth distinguishing from one that could not answer at all — but the
  // caller does the same thing either way, so it is folded into null here.
  if (!summary || typeof summary !== 'object' || !summary.routes) return null;
  return summary;
}
